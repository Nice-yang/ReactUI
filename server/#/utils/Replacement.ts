import * as stable from 'stable';
import * as ts from 'typescript';

export class Replacement {
  public static insert(pos: number, text: string) {
    return new Replacement(pos, pos, text);
  }

  public static delete(start: number, end: number) {
    return new Replacement(start, end, '');
  }

  public static replaceNodeWithText<T extends ts.Node, Y extends ts.Node>(
    source: T,
    node: Y,
    text: string,
  ) {
    return [
      Replacement.delete(
        node.getStart() - source.getStart(),
        node.getEnd() - source.getStart(),
      ),
      Replacement.insert(node.getStart() - source.getStart(), text),
    ];
  }

  public static applyReplacements(source: string, replacements: Replacement[]) {
    replacements = stable(replacements, (r1, r2) => r2.start - r1.start);

    for (const replacement of replacements) {
      source =
        source.slice(0, replacement.start) +
        replacement.text +
        source.slice(replacement.end);
    }

    return source;
  }

  constructor(
    readonly start: number,
    readonly end: number,
    readonly text = '',
  ) {}
}
