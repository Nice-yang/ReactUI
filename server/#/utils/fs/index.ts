import * as fs from 'fs';
import { bindNodeCallback } from 'rxjs';

export const writeFile = bindNodeCallback(fs.writeFile);
