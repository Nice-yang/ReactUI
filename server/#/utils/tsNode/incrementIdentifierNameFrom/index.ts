const incrementName = (name: string) => {
  const regArr = /_\$(\d*)$/.exec(name);

  if (!regArr) {
    return `${name}_$`;
  }

  const n = Number(regArr[1]);

  const index = regArr.index;

  return `${name.slice(0, index)}_$${n + 1}`;
};

const incrementIdentifierNameFrom = (declarationIdentifierNodes: string[]) => {
  const newNames: Record<string, string> = {};

  return (name: string) => {
    if (newNames[name]) {
      return newNames[name];
    }

    let expectedNewName = incrementName(name);

    while (declarationIdentifierNodes.includes(expectedNewName)) {
      expectedNewName = incrementName(expectedNewName);
    }

    newNames[name] = expectedNewName;

    return expectedNewName;
  };
};

export default incrementIdentifierNameFrom;
