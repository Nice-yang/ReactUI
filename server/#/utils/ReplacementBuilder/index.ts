import { Replacement } from '#/utils/Replacement';
import * as ts from 'typescript';

export class ReplacementBuilder {
  private replacements: Replacement[];
  private sourceNode: ts.Node;

  constructor(sourceNode: ts.Node, replacements: Replacement[] = []) {
    this.sourceNode = sourceNode;
    this.replacements = replacements;
  }

  public insert(pos: number, text: string) {
    this.replacements.push(Replacement.insert(pos, text));
  }

  public delete(start: number, end: number) {
    this.replacements.push(Replacement.delete(start, end));
  }

  public deleteNode(node: ts.Node) {
    this.replacements.push(
      Replacement.delete(
        node.getStart() - this.sourceNode.getStart(),
        node.getEnd() - this.sourceNode.getStart(),
      ),
    );
  }

  public replaceNodeWithText(node: ts.Node, text: string) {
    this.replacements = this.replacements.concat([
      Replacement.delete(
        node.getStart() - this.sourceNode.getStart(),
        node.getEnd() - this.sourceNode.getStart(),
      ),
      Replacement.insert(node.getStart() - this.sourceNode.getStart(), text),
    ]);
  }

  public applyReplacements() {
    return Replacement.applyReplacements(
      this.sourceNode.getText(),
      this.replacements,
    );
  }
}
