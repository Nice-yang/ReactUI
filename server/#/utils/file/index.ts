import { Fault } from '#/common/models/Fault';
import { readFile, writeFile } from 'fs';
import * as R from 'ramda';
import { bindNodeCallback, throwError } from 'rxjs';
import { catchError, map } from 'rxjs/operators';

export const readFile$ = bindNodeCallback(readFile);
export const writeFile$ = bindNodeCallback(writeFile);
export const readFileToString$ = (fileName: string) =>
  readFile$(fileName).pipe(
    catchError(err =>
      throwError(
        new Fault('Cannot read file', { error: JSON.stringify(err), fileName }),
      ),
    ),
    map(R.toString),
  );
