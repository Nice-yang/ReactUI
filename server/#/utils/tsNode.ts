import { IJSXSource } from '#/common/models/file';
import * as ts from 'typescript';

export const getInfo = (node: ts.Node) => {
  const file = node.getSourceFile();

  const { line, character: column } = file.getLineAndCharacterOfPosition(
    node.getStart(),
  );

  const fileName = file.fileName;
  const lineNumber = line + 1;
  const columnNumber = column + 1;

  return {
    fileName,
    lineNumber,
    columnNumber,
  };
};

export const isAncestor = (a?: ts.Node, b?: ts.Node): boolean => {
  if (!b || !a) {
    return false;
  }

  if (b.parent === a) {
    return true;
  }

  return isAncestor(a, b.parent);
};

export const findNode = <T extends ts.Node>(
  rootNode: ts.Node,
  target?: IJSXSource,
  predicate?: (node: ts.Node) => boolean,
): T | undefined => {
  let resolvedNode;

  function _traverse(node: ts.Node) {
    const { lineNumber, columnNumber } = getInfo(node);

    if (
      (target &&
        (lineNumber === target.lineNumber &&
          columnNumber === target.columnNumber)) ||
      (predicate && predicate(node))
    ) {
      resolvedNode = node;
      return;
    }

    ts.forEachChild(node, _traverse);
  }

  _traverse(rootNode);

  return resolvedNode;
};

export const findAncestorNode = <T extends ts.Node>(
  rootNode: T,
  predicate: (node: ts.Node) => boolean,
): T | undefined => {
  let resolvedNode;

  let node = rootNode.parent;

  while (node) {
    if (predicate(node)) {
      resolvedNode = node as T;
      break;
    }

    node = node.parent;
  }

  return resolvedNode;
};

export const findAllNodes = <T extends ts.Node>(
  rootNode: ts.Node,
  predicate: (node: ts.Node) => boolean,
): T[] => {
  const resolvedNodes: T[] = [];

  function _traverse(node: ts.Node) {
    if (predicate(node)) {
      resolvedNodes.push(node as T);
    }

    ts.forEachChild(node, _traverse);
  }

  _traverse(rootNode);

  return resolvedNodes;
};
