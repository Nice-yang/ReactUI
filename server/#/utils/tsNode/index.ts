import { Fault } from '#/common/models/Fault';
import { TagCursor } from '#/common/models/file';
import { catchThrowFault } from '#/operators/catchThrowError';
import * as R from 'ramda';
import { Observable } from 'rxjs';
import { pluck, switchMap } from 'rxjs/operators';
import * as ts from 'typescript';
import { createSourceFile$ } from '../tsSourceFile';

export const getInfo = (node: ts.Node) => {
  const file = node.getSourceFile();

  const { line, character: column } = file.getLineAndCharacterOfPosition(
    node.getStart(),
  );

  const fileName = file.fileName;
  const lineNumber = line + 1;
  const columnNumber = column + 1;

  return {
    fileName,
    lineNumber,
    columnNumber,
  };
};

export const findNode$ = <T extends ts.Node>(
  predicate: (node: ts.Node) => boolean,
) => (rootNode: ts.Node): Observable<T> => {
  return new Observable(subscriber => {
    let resolvedNode;

    function _traverse(node: ts.Node) {
      if (predicate(node)) {
        resolvedNode = node as T;
        subscriber.next(resolvedNode);
        subscriber.complete();
        return;
      }

      ts.forEachChild(node, _traverse);
    }

    _traverse(rootNode);

    subscriber.error(new Fault('Node not found', { rootNode }));
  });
};

export const findNodeAtCursor$ = <T extends ts.Node>(
  predicate: (node: ts.Node) => boolean = () => true,
) => (cursor: TagCursor) =>
  createSourceFile$(cursor.fileName).pipe(
    pluck('file'),
    switchMap(findNode$<T>(R.both(isAtCursor(cursor), predicate))),
    catchThrowFault('Node not found at cursor', { cursor }),
  );

export const findElementAtCursor$ = findNodeAtCursor$<ts.JsxElement>(
  ts.isJsxElement,
);

export const findTemplateStringAtCursor$ = findNodeAtCursor$<
  ts.TaggedTemplateExpression
>(ts.isTaggedTemplateExpression);

export const isAtCursor = (cursor: TagCursor) => (node: ts.Node) => {
  const { lineNumber, columnNumber } = getInfo(node);

  return (
    lineNumber === cursor.lineNumber && columnNumber === cursor.columnNumber
  );
};
