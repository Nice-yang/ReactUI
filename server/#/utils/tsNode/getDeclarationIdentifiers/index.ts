import { Declaration, findAllNodes$, isDeclaration } from '#/utils/tsNode';
import { pipe } from 'rxjs';
import { concatMap, toArray } from 'rxjs/operators';
import * as ts from 'typescript';

export const getDeclarationIdentifiersAtNode = (node: Declaration) => {
  if (ts.isVariableDeclaration(node)) {
    return [node.name];
  } else if (ts.isFunctionDeclaration(node)) {
    if (node.name) {
      return [node.name];
    }
  } else {
    if (node.importClause) {
      if (
        node.importClause.namedBindings &&
        ts.isNamespaceImport(node.importClause.namedBindings)
      ) {
        return [node.importClause.namedBindings.name];
      } else {
        if (node.importClause.name) {
          return [node.importClause.name];
        } else if (
          node.importClause.namedBindings &&
          ts.isNamedImports(node.importClause.namedBindings)
        ) {
          return node.importClause.namedBindings.elements.map(
            element => element.name,
          );
        }
      }
    }
  }

  return [];
};

export const getDeclarationIdentifiersAtSourceFile = pipe(
  findAllNodes$<Declaration>(isDeclaration),
  concatMap(getDeclarationIdentifiersAtNode),
  toArray(),
);
