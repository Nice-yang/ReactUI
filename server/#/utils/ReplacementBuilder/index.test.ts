import { findElementAtCursor$ } from '#/utils/tsNode';
import { map, switchMap } from 'rxjs/operators';
import { ReplacementBuilder } from '.';

const fileName = `${__dirname}/__fixtures__/source1.tsx`;

it('should replace div with text', done => {
  findElementAtCursor$({ fileName, lineNumber: 6, columnNumber: 7 })
    .pipe(
      map(cursorNode => {
        const rb = new ReplacementBuilder(cursorNode.getSourceFile());

        rb.replaceNodeWithText(cursorNode, 'replaced text');

        return rb.applyReplacements();
      }),
    )
    .subscribe(fileText => {
      expect(fileText).toMatchSnapshot();
      done();
    });
});

it('should delete node', done => {
  findElementAtCursor$({ fileName, lineNumber: 6, columnNumber: 7 })
    .pipe(
      map(cursorNode => {
        const rb = new ReplacementBuilder(cursorNode.getSourceFile());

        rb.deleteNode(cursorNode);

        return rb.applyReplacements();
      }),
    )
    .subscribe(fileText => {
      expect(fileText).toMatchSnapshot();
      done();
    });
});

it('delete should work with a jsxElement as sourceNode', done => {
  findElementAtCursor$({ fileName, lineNumber: 5, columnNumber: 5 })
    .pipe(
      switchMap(cursorNode => {
        return findElementAtCursor$(
          { fileName, lineNumber: 6, columnNumber: 7 },
          cursorNode,
        ).pipe(
          map(elementNode => {
            const rb = new ReplacementBuilder(cursorNode);

            rb.deleteNode(elementNode);

            return rb.applyReplacements();
          }),
        );
      }),
    )
    .subscribe(nodeText => {
      expect(nodeText).toMatchSnapshot();
      done();
    });
});

it('replaceNodeWithText should work with a jsxElement as sourceNode', done => {
  findElementAtCursor$({ fileName, lineNumber: 5, columnNumber: 5 })
    .pipe(
      switchMap(cursorNode => {
        return findElementAtCursor$(
          { fileName, lineNumber: 6, columnNumber: 7 },
          cursorNode,
        ).pipe(
          map(elementNode => {
            const rb = new ReplacementBuilder(cursorNode);

            rb.replaceNodeWithText(elementNode, 'rrrr');

            return rb.applyReplacements();
          }),
        );
      }),
    )
    .subscribe(nodeText => {
      expect(nodeText).toMatchSnapshot();
      done();
    });
});
