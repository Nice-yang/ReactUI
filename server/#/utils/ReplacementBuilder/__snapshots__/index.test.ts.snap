// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`delete should work with a jsxElement as sourceNode 1`] = `
"<button>
      
      <input value=\\"foo\\" />
    </button>"
`;

exports[`replaceNodeWithText should work with a jsxElement as sourceNode 1`] = `
"<button>
      rrrr
      <input value=\\"foo\\" />
    </button>"
`;

exports[`should delete node 1`] = `
"import * as React from 'react';

const Button = () => {
  return (
    <button>
      
      <input value=\\"foo\\" />
    </button>
  );
};

export default Button;
"
`;

exports[`should replace div with text 1`] = `
"import * as React from 'react';

const Button = () => {
  return (
    <button>
      replaced text
      <input value=\\"foo\\" />
    </button>
  );
};

export default Button;
"
`;
